stdlib/public/core/Rotate.swift

```diff
@@ -0,0 +1,716 @@
+    }
+}
+
+extension RotatedCollection {
+    // An iterator that can be much faster than the iterator of a rotated slice.
+    @_fixed_layout
+    public struct Iterator {
+        @usableFromInline
+        internal var _suffixIterator, _prefixIterator: Base.SubSequence.Iterator
+
+        @inlinable
+        @inline(__always)
+        /// Creates an iterator over the given collection.
+        public /// @testable
+        init(_base: Base, asFirst middle: Base.Index) {
+            _suffixIterator = _base[middle...].makeIterator()
+            _prefixIterator = _base[..<middle].makeIterator()
+        }
+    }
+}
+
+extension RotatedCollection.Iterator: IteratorProtocol {
+    public typealias Element = Base.Element
+
+    @inlinable
+    @inline(__always)
+    public mutating func next() -> Element? {
+        return _suffixIterator.next() ?? _prefixIterator.next()
+    }
+}
+
+extension RotatedCollection: Sequence {
+    /// A type that represents a valid position in the collection.
+    ///
+    /// Valid indices consist of the position of every element and a
+    /// "past the end" position that's not valid for use as a subscript.
+    public typealias Element = Base.Element
+
+    @inlinable
+    @inline(__always)
+    public __consuming func makeIterator() -> Iterator {
+        return Iterator(_base: _base, asFirst: _origin)
+    }
+
+    @inlinable
+    @inline(__always)
+    public var underestimatedCount: Int { return _base.underestimatedCount }
+}
+
+extension RotatedCollection {
+    /// An index that traverses the same positions as an underlying index, in a
+    /// rotated order.
+    @_fixed_layout
+    public struct Index {
+        /// The position of the selected element in the underlying collection,
+        /// or `nil` to indicate the past-the-end position.
+        ///
+        /// To translate in the other direction, pass a `Base.Index` value to
+        /// the `RotatedCollection.rotation(of:)` method that is valid for its
+        /// underlying collection to get a corresponding `Index` value.
+        public let base: Base.Index?
+        /// The position in the underlying collection of the source collection's
+        /// `startIndex`.
+        @usableFromInline
+        internal let _origin: Base.Index
+
+        /// Creates an index for the past-the-end position with the given index
+        /// as the rotated starting index.
+        ///
+        /// - Parameter middle: The source collection's `startIndex`, relative
+        ///   to the underlying collection.
+        ///
+        /// - Complexity: O(1)
+        @inlinable
+        @inline(__always)
+        internal init(startingFrom middle: Base.Index) {
+            base = nil
+            _origin = middle
+        }
+        /// Creates an index into a rotated collection for the given base index,
+        /// with the other given index as the starting position.
+        ///
+        /// - Precondition: `base` cannot be the underlying collection's
```

Comment:
This precondition should be enforced.

Reply:
That `Index` type can't know it.  The precondition has to be maintained from the client side (`RotatedCollection.rotation(of:)`, `RotatedCollection.endIndex`, and their clients within `RotatedCollection`).